### DISCUSSION TITLE
.RPC() returning a error

### DISCUSSION THREAD

**Discussion by HassanIyan** [supabase/supabase#4172] [Questions] [answered]
i try to run .rpc() and this message comes. but in the supabase sql editor it works fine
```
{
  message: 'Could not choose the best candidate function between: public.bookticket(userid => text, tripid => integer, vesselid => integer, qtyofticket => integer), public.bookticket(userid => text, tripid => text, vesselid => text, qtyofticket => integer), public.bookticket(userid => uuid, tripid => integer, vesselid => integer, qtyofticket => integer)',
  hint: 'Try renaming the parameters or the function itself in the database so function overloading can be resolved'
}
```
here is the JS function i am running.
```
let { data, error } = await supabase
    .rpc('bookticket', {
      userid: 'c7ae6b9e-c83e-4847-9f6c-0f0ecc4452af',
      tripid: 4,
      vesselid: 3,
      qtyofticket: 10
    })

  if (error) console.error(error)
  else console.log(data)
```
here is the plpgsql code
```
create or replace function bookticket(userid text, tripid integer, vesselid integer, qtyofticket integer) 
returns JSON
language plpgsql
as $$
declare
bookingid integer;
begin
  
  
  IF (SELECT available_seats > qtyofticket FROM trips WHERE id = tripid) THEN
    INSERT into bookings(trip_id, vessel_id, validity, user_id, qty, paid) VALUES(tripid, vesselid, true, userid, qtyofticket, false) RETURNING id into bookingid;
    update trips set available_seats = available_seats - qtyofticket where id = tripid;
    return json_build_object('msg', bookingid);
  ELSE
    return json_build_object('msg', 'there are no available seats');
  END IF;
  

END;
$$;
```

**Comment by GaryAustin1** [comment, marked as answer]
You changed the parameter type of the function at some point which created 2 functions.   Function name and parameter name and parameter type are all part of identifying the correct function.  You need to delete the old function(s)  or as a quick test create a new function name and it should work. 

From Postgres manual:
"If a schema name is included, then the function is created in the specified schema. Otherwise it is created in the current schema. The name of the new function must not match any existing function with the same input argument types in the same schema. However, functions of different argument types can share a name (this is called overloading).

To replace the current definition of an existing function, use CREATE OR REPLACE FUNCTION. It is not possible to change the name or argument types of a function this way (if you tried, you would actually be creating a new, distinct function). Also, CREATE OR REPLACE FUNCTION will not let you change the return type of an existing function. To do that, you must drop and recreate the function. (When using OUT parameters, that means you cannot change the types of any OUT parameters except by dropping the function.)"

See this: https://www.postgresql.org/docs/13/sql-dropfunction.html to help clean things up...

**Reply by steve-chavez** [reply]
What Gary mentions is true, this is related to overloading. 

However this error comes from PostgREST as its support for [function overloading](https://postgrest.org/en/v9.0/api.html#overloaded-functions) is not the same as PostgreSQL(this is why calling the function in the SQL editor works).

> Important
> Overloaded functions with the same argument names but different types are not supported.

In the error message we can see that there are 3 functions with the same parameter names but with different types:

```
public.bookticket(userid => text, tripid => integer, vesselid => integer, qtyofticket => integer), 
public.bookticket(userid => text, tripid => text, vesselid => text, qtyofticket => integer), 
public.bookticket(userid => uuid, tripid => integer, vesselid => integer, qtyofticket => integer)
```

It's likely that a couple of those functions are unused, dropping them would solve the issue.

---

Btw, sorry for the confusion caused, we'll differentiate between PostgREST and PostgreSQL errors(https://github.com/PostgREST/postgrest/issues/1917) on the next release. PostgREST strives to be fully transparent to PostgreSQL but there are some few areas where we need to add additional logic.
